### PR TITLE
Detect if the background processor exits early and shutdown if so

### PR DESCRIPTION
**Pull request by valentinewallace**
A user ran into an issue where the `BackgroundProcessor` silently stopped, leading to `ChannelManager` persistence falling behind + channel force closures. Now we'll detect if the bg processor exits early and log the result. 

Still need to do a bit more reading up on tokio docs to vet the patch but feedback welcome.

### CODE REVIEW

File: src/main.rs [hunk 1]
```diff
@@ -897,21 +897,26 @@ async fn start_ldk() {
...
+			Arc::clone(&channel_manager),
+			Arc::clone(&keys_manager),
+			Arc::clone(&network_graph),
+			Arc::clone(&onion_messenger),
+			inbound_payments,
+			outbound_payments,
+			ldk_data_dir,
+			network,
+			Arc::clone(&logger),
+			Arc::clone(&persister),
```
Comment (jkczyz) [inline src/main.rs:914]: Could you make this a separate variable?

File: src/main.rs [hunk 2]
```diff
@@ -920,7 +925,6 @@ async fn start_ldk() {
 
 	// Stop the background processor.
 	bp_exit.send(()).unwrap();
```
Comment (jkczyz) [inline src/main.rs:927]: Will this work if the receiver has been dropped? Maybe condition this on `!bp_exit.is_closed()`.

File: src/main.rs [hunk 3]
```diff
@@ -897,21 +897,26 @@ async fn start_ldk() {
...
+			Arc::clone(&onion_messenger),
+			inbound_payments,
+			outbound_payments,
+			ldk_data_dir,
+			network,
+			Arc::clone(&logger),
+			Arc::clone(&persister),
+		) => {},
+		bg_res = background_processor => {
+			println!("ERR: background processing stopped with result {:?}, exiting", bg_res);
```
Comment (jkczyz) [inline src/main.rs:917]: We should either panic here or make sure the process exits with an error code.

File: src/main.rs [hunk 4]
```diff
@@ -897,21 +897,26 @@ async fn start_ldk() {
...
-		ldk_data_dir,
-		network,
-		Arc::clone(&logger),
-		Arc::clone(&persister),
-	)
-	.await;
+	// Exit if CLI polling exits or the background processor exits (which shouldn't happen unless we
+	// fail to write to the filesystem).
+	tokio::select! {
+		_ = cli::poll_for_user_input(
```
Comment (jkczyz) [inline src/main.rs:903]: Not sure if it is strictly required, but you may want to spawn this instead.

https://docs.rs/tokio/latest/tokio/macro.select.html#runtime-characteristics

Comment (valentinewallace) [reply src/main.rs]: Based on those docs, I can't tell if with the way I had it written previously we would have gotten stuck on the CLI `select` branch. LMK if you know, for my understanding.

Comment (jkczyz) [reply src/main.rs]: Probably not stuck given the background processor task was already spawned and would have been running that future already.

https://docs.rs/tokio/latest/tokio/task/fn.spawn.html

Comment (valentinewallace) [reply src/main.rs]: Ah, I meant whether we would ever hit the other `select` branch to check for the bg processor exiting. Looks like it does work: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=0c2f28caa7c80e47d969465ba17e9986

**Review by jkczyz — APPROVED** [review]

**Pull request merged**